### PR TITLE
[Accelerator 4] Adding real time event notification

### PR DESCRIPTION
**Pull request by Ashi1993**
## Adding real time event notification

> This PR adds Realtime Event Notification Implementation

**Issue link:** *https://github.com/wso2-enterprise/solutions-bfsi/issues/20*

**Doc Issue:** *Optional, link issue from [documentation repository](https://github.com/wso2/docs-ob/issues)*

**Applicable Labels:** *Spec, product, version, type (specify requested labels)*

------

### Development Checklist

1. [x] Build complete solution with pull request in place.
2. [x] Ran checkstyle plugin with pull request in place.
3. [x] Ran Findbugs plugin with pull request in place.
4. [x] Ran FindSecurityBugs plugin and verified report.
5. [x] Formatted code according to WSO2 code style.
6. [x] Have you verified the PR doesn't commit any keys, passwords, tokens, usernames, or other secrets?
7. [ ] Migration scripts written (if applicable).
8. [x] Have you followed secure coding standards in [WSO2 Secure Engineering Guidelines](http://wso2.com/technical-reports/wso2-secure-engineering-guidelines)?

### Testing Checklist

1. [ ] Written unit tests.
2. [ ] Verified tests in multiple database environments (if applicable).
3. [ ] Tested with BI enabled  (if applicable).

### CODE REVIEW

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/pom.xml [hunk 1]
```diff
@@ -204,8 +204,9 @@
                             org.wso2.carbon.identity.oauth2.*;version="${identity.inbound.auth.oauth.version.range}",
                             org.wso2.financial.services.accelerator.common.*;version="${project.version}",
                             org.wso2.financial.services.accelerator.consent.mgt.dao.*;version="${project.version}",
-                            org.wso2.financial.services.accelerator.consent.mgt.service.*;version="${project.version}",
+                            org.wso2.financial.services.accelerator.consent.mgt.service.*;version="${project.version}"
                         </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
Comment (Akila94) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/pom.xml:209]: As discussed offline, let's create an issue to remove this dynamic imports line in the future.

Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/pom.xml:209]: Created issue https://github.com/wso2/financial-services-accelerator/issues/218

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/RealtimeNotificationService.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            throw new FSEventNotificationException(EventNotificationConstants.ERROR_STORING_EVENT_SUBSCRIPTION, e);
+        } finally {
+            log.debug(EventNotificationConstants.DATABASE_CONNECTION_CLOSE_LOG_MSG);
+            DatabaseUtils.closeConnection(connection);
+        }
+    }
+
+    /**
+     * Method to update the notification status by ID, allowed values are.
+     * OPEN,ACK and ERR
```
Comment (aka4rKO) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/RealtimeNotificationService.java:111]: if only these values are allowed shouldn't we accept an enum?

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/realtime/service/RealtimeEventNotificationSenderService.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                                + notificationId.replaceAll("[\r\n]", ""));
+                    }
+                }
+                retryCount++;
+
+                // If the circuit breaker is opened or the maximum retry count is exceeded,
+                // the notification status will be updated as ERROR.
+                realtimeNotificationService.updateNotificationStatusById(notificationId,
+                        EventNotificationConstants.ERROR);
+            } catch (IOException | InterruptedException  e) {
```
Comment (aka4rKO) [inline financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/realtime/service/RealtimeEventNotificationSenderService.java:195]: if these both exception are logging the same error we can merge them

**Review by RivinduM — APPROVED** [review]

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/realtime/service/RealtimeEventNotificationSenderService.java (hunk 3 above)
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/realtime/service/RealtimeEventNotificationSenderService.java]: Fixed in https://github.com/wso2/financial-services-accelerator/pull/193/commits/00babc11f6c634ec7023e46cba1a28b73d4377e7

File: financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/RealtimeNotificationService.java (hunk 2 above)
Comment (Ashi1993) [reply financial-services-accelerator/components/org.wso2.financial.services.accelerator.event.notifications.service/src/main/java/org/wso2/financial/services/accelerator/event/notifications/service/RealtimeNotificationService.java:111]: FIxed in https://github.com/wso2/financial-services-accelerator/pull/193/commits/00babc11f6c634ec7023e46cba1a28b73d4377e7

**Review by aka4rKO — APPROVED** [review]

**Comment by Ashi1993** [comment]
Closing this PR and added this again because https://github.com/wso2/financial-services-accelerator/pull/224 due to an issue with CLA.

**Pull request closed without merging**